cpp/include/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,489 @@
+ * @tparam Rad The radix/base (either `Radix::BASE_2` or `Radix::BASE_10`)
+ *---------------------------------------------------------------------------**/
+template <typename Rep, Radix Rad>
+class fixed_point {
+
+    Rep        _value;
+    scale_type _scale;
+
+public:
+
+    /**---------------------------------------------------------------------------*
+    * @brief Constructor that will perform shifting to store value appropriately
+    *
+    * @tparam T The type that you are constructing from (integral or floating)
+    * @param value The value that will be constructed from
+    * @param scale The exponent that is applied to Rad to perform shifting
+    *---------------------------------------------------------------------------**/
+    template <typename T,
+              typename std::enable_if_t<is_supported_construction_value_type<T>()
+                                     && is_supported_representation_type<Rep>()>* = nullptr>
+    explicit fixed_point(T const& value, scale_type const& scale) :
+        _value{static_cast<Rep>(detail::shift_with_precise_round<Rep, Rad>(value, scale))},
+        _scale{scale}
+    {
+    }
+
+    /**---------------------------------------------------------------------------*
+    * @brief Constructor that will not perform shifting (assumes value already
+    * shifted)
+    *
+    * @param s scaled_integer that contains scale and already shifted value
+    *---------------------------------------------------------------------------**/
+    explicit fixed_point(scaled_integer<Rep> s) :
+        _value{s.value},
+        _scale{s.scale}
+    {
+    }
+
+   /**---------------------------------------------------------------------------*
+    * @brief Default onstructor that constructs `fixed_point` number with a
+    * value and scale of zero
+    *---------------------------------------------------------------------------**/
+    fixed_point() :
+        _value{0},
+        _scale{scale_type{0}}
+    {
+    }
+
+   /**---------------------------------------------------------------------------*
+    * @brief Explicit conversion operator
+    *
+    * @tparam U The type that is being explicitly converted to (integral or floating)
+    * return The `fixed_point` number in base 10 (aka human readable format)
+    *---------------------------------------------------------------------------**/
+    template <typename U,
+              typename std::enable_if_t<is_supported_construction_value_type<U>()>* = nullptr>
+    CUDA_HOST_DEVICE_CALLABLE
+    explicit constexpr operator U() const {
+        return detail::shift<Rep, Rad>(static_cast<U>(_value), detail::negate(_scale));
+    }
+
+    /**---------------------------------------------------------------------------*
+    * @brief operator +=
+    *
+    * @tparam Rep1 Representation type of number being added to `this`
+    * @tparam Rad1 Radix (base) type of number being added to `this`
+    * @return Modified `this`
```

Comment:
I don't think "Modified `this`" is helpful. 
```suggestion
    * @return The sum.
```